Snail/Disposer.swift

```diff
@@ -7,16 +7,25 @@ public protocol DisposableType {
 }
 
 public class Disposer {
-    private(set) public var disposables: [DisposableType] = []
+    private(set) var disposables: [DisposableType] = []
+    private let recursiveLock = NSRecursiveLock()
 
     public init() {}
 
+    deinit {
+        disposeAll()
+    }
+
     public func disposeAll() {
+        recursiveLock.lock()
         disposables.forEach { $0.dispose() }
         disposables.removeAll()
+        recursiveLock.unlock()
     }
 
     public func add(disposable: DisposableType) {
+        recursiveLock.lock()
         disposables.append(disposable)
+        recursiveLock.unlock()
```

Comment:
One final thing: Id update all usage of lock to do:

`lock.lock; defer { lock.unlock }`

It prevents other people from accidentally adding code at the end of the unlock